### PR TITLE
Twilio receive text message agent

### PR DESCRIPTION
**Pull request by albertsun**
To be merged after https://github.com/cantino/huginn/pull/1415

An agent for receiving text messages from Twilio. It requires somewhat different logic from the existing webhook agent.

### CODE REVIEW

**Comment by cantino** [comment]
Cool!  Do you think it would make sense to merge this into the existing Twilio Agent?  I'm generally a fan of fewer agents.

**Comment by albertsun** [comment]
If this were part of the existing Twilio agent it would need a `mode` option or something to switch between read/write. To me that's more confusing than simply having separate agents for sending to and receiving from Twilio. (Kind of the way there are different agents for different Twitter actions).

**Comment by cantino** [comment]
@dsander / @knu, this is coming up a lot.  Do either of you have an opinion on unified Agents vs Agent-per-action?

**Comment by dsander** [comment]
I generally prefer a couple simple agents over one complicated one that does a lot, especially when the separation can be made clear in the agent name/description. While writing the LocalFile and CsvAgent I was unsure and went with the `mode` approach but regretted it relatively soon, the code and description got pretty complicated.

We could introduce a general `mode` concept to agents which would dynamically change the description, the from configurable fields or the default options.

**Comment by albertsun** [comment]
If the goal is to have a more navigable number of agents on the New Agent screen, what about grouping agents into related modules like `Agents::Twilio::SendText` and `Agents::Twilio::ReceiveText`. They could then be grouped together on the New screen and in other places.

**Comment by cantino** [comment]
That seems like a good idea @albertsun.  I think I'm still biased towards just having a TwilioAgent that can do a couple things (e.g., one agent per integration instead of one agent per integration behavior), but I don't feel strongly.

File: app/models/agent.rb [hunk 1]
```diff
@@ -149,15 +153,21 @@ def update_event_expirations!
...
       receive_webhook(params).tap do
         self.last_web_request_at = Time.now
         save!
       end
     else
-      receive_web_request(params, method, format).tap do
+      if method(:receive_web_request).arity == 1
+        handled_request = receive_web_request(request)
+      else
+        handled_request = receive_web_request(params, request.method_symbol.to_s, request.format)
```
Comment (cantino) [inline app/models/agent.rb]: Do we need `request.format.to_s` to match what was in the controller?

**Comment by albertsun** [comment]
Now that https://github.com/cantino/huginn/pull/1415 is merged I rebased this on master so there's just one commit.

As for one agent that does many things vs many agents doing one thing each, I took a look through the @dsander 's CsvAgent and LocalFileAgent and the code around "mode" seems a little more complicated than I like. It's almost like writing two agents in one with conditionals and feels like it'll have harder to use tests and documentation than having simpler agents. Maybe if some of the `mode` concepts were implemented at the agent level?

Either way, wondering if we can just merge this agent, and discuss that further in a broader issue

File: app/models/agents/twilio_receive_text_agent.rb [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+      #{'## Include `twilio-ruby` in your Gemfile to use this Agent!' if dependencies_missing?}
+
+      In order to create events with this agent, configure Twilio to send POST requests to:
+
+      ```
+      #{post_url}
+      ```
+
+      #{'The placeholder symbols above will be replaced by their values once the agent is saved.' unless id}
```
Comment (cantino) [inline app/models/agents/twilio_receive_text_agent.rb]: Nice :)

File: app/models/agents/twilio_receive_text_agent.rb [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+      Options:
+
+        * `server_url` must be set to the URL of your
+        Huginn installation (probably "https://#{ENV['DOMAIN']}"), which must be web-accessible.  Be sure to set http/https correctly.
+
+        * `account_sid` and `auth_token` are your Twilio account credentials. `auth_token` must be the primary auth token for your Twilio accout.
```
Comment (cantino) [inline app/models/agents/twilio_receive_text_agent.rb]: Do they support oAuth?

File: app/models/agents/twilio_receive_text_agent.rb [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    end
+
+    def receive_web_request(request)
+      params = request.params.except(:action, :controller, :agent_id, :user_id, :format)
+      method = request.method_symbol.to_s
+      headers = request.headers
+      
+      # check the last url param: 'secret'
+      secret = params.delete('secret')
+      return ["Not Authorized", 401] unless secret == "sms-endpoint"
```
Comment (cantino) [inline app/models/agents/twilio_receive_text_agent.rb]: If we drop this and make the route just end in `:id/`, does it work?  Does having this check add any value?

Comment (cantino) [inline app/models/agents/twilio_receive_text_agent.rb]: Never mind, I recall a secret being required.

File: app/models/agents/twilio_receive_text_agent.rb [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            r.Message interpolated['reply_text']
+          end
+        end
+        return [response.text, 201, "text/xml"]
+      else
+        return ["Bad request", 400]
+      end
+    end
+
+    # def client
```
Comment (cantino) [inline app/models/agents/twilio_receive_text_agent.rb]: Removable?

**Comment by cantino** [comment]
Agreed, I don't want to block this on the larger discussion about `mode` and one-vs-many agents.

File: app/models/agents/twilio_receive_text_agent.rb (hunk 5 above)
Comment (albertsun) [inline app/models/agents/twilio_receive_text_agent.rb]: Yep removeable

File: app/models/agents/twilio_receive_text_agent.rb (hunk 3 above)
Comment (albertsun) [inline app/models/agents/twilio_receive_text_agent.rb]: They don't support oauth for verifying message signatures

**Comment by albertsun** [comment]
Comment removed. Hope this is ready to merge now.

(As for mode and one vs many discussion, perhaps that is also part of the pulling agents into gems discussion — there could be a huginn-twilio gem that has multiple `Agents::Twilio::FooAgent` classes)

**Comment by cantino** [comment]
Nice work!

**Pull request merged**